src/client/core/utils/dom.js

```diff
@@ -364,28 +367,78 @@ export function getElementDescription (el) {
+    const isArrowNavigationBetweenNonamesDisallowed = !skipRadioGroups && !sourceElement.name && !browserUtils.isChrome;
+
+    elements = arrayUtils.filter(elements, item => {
+        const isRadioItem = isRadioButtonElement(item);
+
+        if (isArrowNavigationAllowed)
+            return isRadioItem && item.name === sourceElement.name;
+
+        if (isArrowNavigationBetweenNonamesAllowed)
+            return isRadioItem && !item.name;
+
+        if (isArrowNavigationBetweenNonamesDisallowed)
+            return item === sourceElement;
+
+        // NOTE: in all browsers except Mozilla and Opera focus sets on one radio set from group only.
+        // in Mozilla and Opera focus sets on any radio set.
+        if (sourceElement.name !== '' && !browserUtils.isFirefox)
+            return !item.name || item === sourceElement || item.name !== sourceElement.name;
+
+        return true;
+    });
+
+    return elements;
+}
+
+function correctFocusableElement (elements, element, skipRadioGroups) {
+    const isNotCheckedRadioButtonElement = isRadioButtonElement(element) && element.name && !element.checked;
+
+    if (!skipRadioGroups || !isNotCheckedRadioButtonElement)
+        return element;
+
+    const checkedRadioButtonElementWithSameName = arrayUtils.find(elements, el => {
+        return isRadioButtonElement(el) && el.name === element.name && el.checked;
+    });
+
+    return checkedRadioButtonElementWithSameName || element;
+}
+
+export function getNextFocusableElement (element, reverse, skipRadioGroups) {
+    const offset     = reverse ? -1 : 1;
     let allFocusable = sortElementsByFocusingIndex(getFocusableElements(findDocument(element)));
 
-    //NOTE: in all browsers except Mozilla and Opera focus sets on one radio set from group only.
-    // in Mozilla and Opera focus sets on any radio set.
-    if (isInputElement(element) && element.type === 'radio' && element.name !== '' && !browserUtils.isFirefox) {
-        allFocusable = arrayUtils.filter(allFocusable, item => {
-            return !item.name || item === element || item.name !== element.name;
-        });
-    }
+    allFocusable = filterFocusableElements(allFocusable, element, skipRadioGroups);
 
+    const isRadioInput         = isRadioButtonElement(element);
     const currentIndex         = arrayUtils.indexOf(allFocusable, element);
     const isLastElementFocused = reverse ? currentIndex === 0 : currentIndex === allFocusable.length - 1;
 
     if (isLastElementFocused)
-        return document.body;
+        return skipRadioGroups || !isRadioInput ? document.body : allFocusable[allFocusable.length - 1 - currentIndex];
 
     if (reverse && currentIndex === -1)
         return allFocusable[allFocusable.length - 1];
 
-    return allFocusable[currentIndex + offset];
+    return correctFocusableElement(allFocusable, allFocusable[currentIndex + offset], skipRadioGroups);
```

Comment:
What if `currentIndex === -1`? Also looks like `correctFocusableElement` doesn't take `reverse` into account.

Reply:
`correctFocusableElement` always returns radio button with specific name and it does not depend on order

Reply:
even if currentIndex === -1, there will not be an error because it's corrected by `offset`.